packages/orm/src/EntityManager.ts

```diff
@@ -482,30 +492,48 @@ export class EntityManager {
    * (i.e. including the initial `SELECT`s) in a transaction.
    */
   async flush(): Promise<void> {
-    if (this.currentFlushPromise) {
-      await this.currentFlushPromise;
-      await this.flushLoader.load(0);
-    } else {
-      await this.flushLoader.load(0);
+    if (this.isFlushing) {
+      throw new Error("Cannot flush while another flush is already in progress");
     }
+
+    await this.flushLoader.load(0);
+  }
+
+  get isFlushing(): boolean {
+    return this.currentFlushPromise !== undefined;
   }
 
   /** The implementation of flush, but called by a DataLoader to de-dup calls made in a loop. */
   private async doFlush(): Promise<void> {
-    // We defer doing this cascade logic until flush() so that delete() can remain synchronous.
-    await this.cascadeDeletesIntoUnloadedCollections();
-    recalcDerivedFields(this.entities);
-    const entityTodos = sortEntities(this.entities);
-    await addReactiveAsyncDerivedValues(entityTodos);
-    await recalcAsyncDerivedFields(this, entityTodos);
+    const entitiesToFlush: Entity[] = [];
+    let pendingEntities = this.entities.filter((e) => e.isPendingFlush);
+
+    while (pendingEntities.length > 0) {
```

Comment:
Eventually could put in a sanity check counter like `numberOfLoops` and if it gets over 5, console.log a warn, and if it gets over say 10 or 100, blow up.